content/blog/2019-11-13-efficient-path-prof/index.md

```diff
@@ -0,0 +1,205 @@
+| [FloatMM](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/Stanford/FloatMM.c) | 1.48 | 3.61 | 2.44 |
+| [huffbench](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/CoyoteBench/huffbench.c) | 55.67 | 102.73 | 1.84|
+| [linpack-pc](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/Linpack/linpack-pc.c) | 15.80 | 37.49 | 2.37 |
+| [lpbench](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/CoyoteBench/lpbench.c)  | 10.94 | 26.75 | 2.45 |
+| [misr](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/McGill/misr.c) | 0.32 | 0.41 | 1.28|
+| [oourafft](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/Misc/oourafft.c) | 13.45 | 18.24 | 1.36 |
+| [queens](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/McGill/queens.c) | 3.53 | 6.69 | 1.90 |
+| [ReedSolomon](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/Misc/ReedSolomon.c) | 11.28 | 19.35 | 1.72 |
+| [Treesort](https://github.com/llvm/llvm-test-suite/blob/master/SingleSource/Benchmarks/Stanford/Treesort.c) | 0.18 | 0.32 |1.78 |
+
+
+These results were acquired in one run of each program. There was no effort made to isolate the running process from other process on the system (a standard laptop) or efforts to minimize the effects of memory alignment. This was because the performance differences were large enough that doing any of the above would have likely been overkill to get meaningful results.
+
+The results show worse overhead than the paper showed, but this may be the result of multiple factors. The first is that the algorithm that I implemented did not include any of the instrumentation placement optimizations that the paper described. In my implementation, the path register is actually stack allocated so every edge instrumentation needs to perform and memory read and write. Instead, the path register can actually be pinned to a register throughout a function. This would cause one less register to be available for register allocation so it would be interesting to see what the tradeoffs would be. Next, the operation to increment the path counter table is actually a function call. This is because I thought I would add locking to support multithreaded programs. However, none of the benchmarks are multithreaded, so this function call is unecessary and the increment could be inlined. Finally, there is no strength reduction to store the address of path counter in the path register instead of the index. The paper mentions this as a potential optimization, however no optimzation passes were run on the programs after instrumentation.
+
+In summary, the instrumentation generally causes about a 2x increase in program execution time, which is not that great. It would be interesting to see what performance improvments any or all of the above optimizations could achieve.
```

Comment:
It's not so bad! I would bet that your overheads would look a lot smaller if you optimized both the baseline and your instrumented code with `-O2`, for example.